File: BaeKit/Sources/BaeKit/Services/AppService.swift
import Combine
import Foundation
import Observation
import SwiftUI

/// The shared services one platform `AppService` constructs before handing
/// ownership to its BaeKit base. Its fields are deliberately hidden: the value
/// crosses the superclass-initialization boundary, but none of the retained
/// services can be unpacked afterward.
public struct AppServiceComponents {
    fileprivate let playbackStore: PlaybackStore
    fileprivate let configStore: ConfigStore
    fileprivate let syncStatusStore: SyncStatusStore
    fileprivate let artworkLoadingStore: ArtworkLoadingStore
    fileprivate let libraryStore: LibraryStore
    fileprivate let downloadStore: DownloadStore
    fileprivate let castStore: CastStore
    fileprivate let outboxStore: OutboxStore
    fileprivate let library: Library
    fileprivate let playback: Playback
    fileprivate let queue: Queue
    fileprivate let mediaPaths: MediaPaths
    fileprivate let imageStore: ImageStore
    fileprivate let sync: Sync
    fileprivate let downloads: Downloads
    fileprivate let cast: Cast

    public init(
        playbackStore: PlaybackStore,
        configStore: ConfigStore,
        syncStatusStore: SyncStatusStore,
        artworkLoadingStore: ArtworkLoadingStore,
        libraryStore: LibraryStore,
        downloadStore: DownloadStore,
        castStore: CastStore,
        outboxStore: OutboxStore,
        library: Library,
        playback: Playback,
        queue: Queue,
        mediaPaths: MediaPaths,
        imageStore: ImageStore,
        sync: Sync,
        downloads: Downloads,
        cast: Cast
    ) {
        self.playbackStore = playbackStore
        self.configStore = configStore
        self.syncStatusStore = syncStatusStore
        self.artworkLoadingStore = artworkLoadingStore
        self.libraryStore = libraryStore
        self.downloadStore = downloadStore
        self.castStore = castStore
        self.outboxStore = outboxStore
        self.library = library
        self.playback = playback
        self.queue = queue
        self.mediaPaths = mediaPaths
        self.imageStore = imageStore
        self.sync = sync
        self.downloads = downloads
        self.cast = cast
    }
}

#if DEBUG
    public struct AppServiceTestState {
        public let nowPlaying: NowPlaying
        public let playbackPosition: PlaybackPositionEvent
        public let manualQueueEntryIds: [String]
        public let upcomingQueueEntryIds: [String]?
        public let volume: Float
        public let isMuted: Bool
        public let repeatMode: BridgeRepeatMode
    }
#endif

/// Root application object for one unlocked library — the platform-shared base.
/// Not `@Observable`: the reactive state lives on the contained stores. Conforms
/// to `Observable` (empty) so it can sit in SwiftUI's `@Environment`.
///
/// Each platform subclasses this with its own stored services (the desktop
/// import/export layer on macOS) and its own `init` / `wireUp`; the shared
/// stores, domain services, common projections, and media-control plumbing live
/// here. One instance per unlocked library: switching libraries builds a fresh
/// one with fresh stores.
@MainActor
open class AppService: @unchecked Sendable, Observable {
    // The immutable `Sendable` handle and domain services are marked
    // `nonisolated` so they read from any context (the app delegates touch
    // `appHandle`/`sync` off the main actor when tearing a library down). Within
    // a module the compiler infers this for `let`s of `Sendable` type on a
    // `@MainActor` class; across the package boundary it must be explicit.

    /// Handle to bae-core via the uniffi bridge. All library, playback, and
    /// sync requests go through this.
    private nonisolated let appHandle: AppHandle

    /// The process-lifetime telemetry sink (built at launch, before this
    /// library opened). Host-originated events route through it; unlike
    /// `appHandle` it outlives any one library, so switching libraries keeps the
    /// same sink.
    private nonisolated let diagnostics: BridgeDiagnostics

    /// Drives the platform Now Playing surface and remote-control events, kept
    /// in sync with `playbackStore.nowPlaying`.
    private nonisolated let mediaControlService: MediaControlService

    // MARK: - Stores

    private let playbackStore: PlaybackStore
    private let configStore: ConfigStore
    private let syncStatusStore: SyncStatusStore
    private let artworkLoadingStore: ArtworkLoadingStore
    private let libraryStore: LibraryStore
    /// In-memory download (pin) queue mirror — per-release state and summary.
    private let downloadStore: DownloadStore
    /// Which device playback is on, and what the picker found while it was open.
    private let castStore: CastStore
    /// Cloud outbox processing mirror — queue depth, per-item state, summary
    /// (carries the sync-pause flag).
    private let outboxStore: OutboxStore

    private let commonSubscriptions: CommonSubscriptions
    private let libraryProjectionStore: LibraryProjectionStore
    private lazy var libraryListsStore = LibraryListsStore(
        library: library,
        libraryStore: libraryStore,
        onError: { [weak self] error in self?.showError(error) }
    )

    // MARK: - Domain services
    //
    // Narrow, per-domain projections of `AppHandle` that views read via
    // `@Environment(...)` instead of reaching for `appService.appHandle`.

    private nonisolated let library: Library
    private nonisolated let playback: Playback
    private nonisolated let queue: Queue
    private nonisolated let mediaPaths: MediaPaths
    /// The app's image pipeline: fetch, decode-at-size, bounded decoded cache.
    private nonisolated let imageStore: ImageStore
    private nonisolated let sync: Sync
    private nonisolated let downloads: Downloads
    /// Cast transport: browse for devices, and move playback to or from one.
    private nonisolated let cast: Cast
    private let playbackValues: PlaybackValueHandler

    #if DEBUG
        private let testAccess: AppServiceTestAccess
    #endif

    /// Build the shared half of an `AppService` around an already-open,
    /// already-unlocked `AppHandle` and the services constructed for it. Pure
    /// field assignment; every side effect lives in the subclass `wireUp()`
    /// after construction.
    public init(
        appHandle: AppHandle,
        mediaControlService: MediaControlService,
        diagnostics: BridgeDiagnostics,
        components: AppServiceComponents
    ) {
        self.appHandle = appHandle
        self.mediaControlService = mediaControlService
        self.diagnostics = diagnostics
        playbackStore = components.playbackStore
        configStore = components.configStore
        syncStatusStore = components.syncStatusStore
        artworkLoadingStore = components.artworkLoadingStore
        libraryStore = components.libraryStore
        downloadStore = components.downloadStore
        castStore = components.castStore
        outboxStore = components.outboxStore
        commonSubscriptions = CommonSubscriptions(
            appHandle: appHandle,
            configStore: components.configStore,
            syncStatusStore: components.syncStatusStore,
            artworkLoadingStore: components.artworkLoadingStore,
            outboxStore: components.outboxStore,
            downloadStore: components.downloadStore,
            castStore: components.castStore
        )
        libraryProjectionStore = LibraryProjectionStore(
            library: components.library
        )
        library = components.library
        playback = components.playback
        queue = components.queue
        mediaPaths = components.mediaPaths
        imageStore = components.imageStore
        sync = components.sync
        downloads = components.downloads
        cast = components.cast
        playbackValues = PlaybackValueHandler(
            playbackStore: components.playbackStore,
            castStore: components.castStore
        )
        #if DEBUG
            testAccess = AppServiceTestAccess(
                playbackStore: components.playbackStore
            )
        #endif
    }

    public var libraryId: String { configStore.config.libraryId }
    public var libraryName: String { configStore.config.libraryName }
    public var libraryPath: String { configStore.config.libraryPath }

    public var playbackPositionPublisher:
        AnyPublisher<PlaybackPositionEvent, Never>
    {
        playbackStore.playbackPositionPublisher
    }

    public func savePlaybackState() async throws {
        try await appHandle.savePlaybackState()
    }

    public nonisolated func forgetLibrary() async throws {
        try await appHandle.forgetLibrary()
    }

    public func shutdown() async throws {
        try await appHandle.shutdown()
    }

    public func triggerSync() {
        sync.triggerSync()
    }

    public nonisolated func renameLibrary(
        _ libraryId: String,
        to name: String
    ) throws {
        try sync.renameLibrary(libraryId, name)
    }

    public nonisolated func lockActiveLibrary() async throws {
        try await sync.lockActiveLibrary()
    }

    public func storeRestoreCodeInKeychain(
        libraryId: String,
        onError: @escaping @Sendable (DisplayError) -> Void
    ) {
        sync.storeRestoreCodeInKeychain(
            libraryId: libraryId,
            onError: onError
        )
    }

    #if os(iOS)
        public func setupRemoteCommands() {
            mediaControlService.setupRemoteCommands(
                playback: playback,
                playbackStore: playbackStore
            )
        }
    #else
        public func activateMediaControls(previewAudio: PreviewAudio) {
            mediaControlService.activate(
                playback: playback,
                previewAudio: previewAudio,
                playbackStore: playbackStore
            )
        }
    #endif

    #if os(macOS)
        public func deactivateMediaControls() {
            mediaControlService.deactivate(playbackStore: playbackStore)
        }

        public func registerArtworkAnalyzer(
            _ analyzer: ArtworkAnalyzerCallback
        ) {
            appHandle.registerArtworkAnalyzer(analyzer: analyzer)
        }
    #endif

    public func subscribeUIEvents(
        onUnhandled:
            @escaping @MainActor @Sendable (BridgeUiEvent) -> Void
    ) {
        appHandle.subscribeUiEvents(
            callback: UiEventPump(
                sink: UiEventDispatcher.makeSink(
                    appService: self,
                    onUnhandled: { event, _ in onUnhandled(event) }
                )
            )
        )
    }

    /// Report that a host UI screen was opened, as a typed telemetry event,
    /// through the process-lifetime telemetry sink. Infallible — telemetry must
    /// never affect navigation. This is the only host-originated event;
    /// everything else the core emits itself. Local logging is separate
    /// (`BaeLogger`) and stays on-device.
    public nonisolated func reportScreen(_ screen: BridgeScreen) {
        diagnostics.event(event: .screenOpened(screen: screen))
    }

    /// Route a caught error to the platform's error surface. An error core says
    /// has no line — a cancellation — is dropped rather than surfaced empty.
    public func showError(_ error: any Error) {
        guard let displayed = DisplayError(error) else { return }
        showError(displayed)
    }

    /// Route a playback failure. It is not a Swift `Error`, so it needs its own
    /// door; it drops on a nil line for the same reason.
    public func showError(_ reason: BridgePlaybackErrorReason) {
        guard let displayed = DisplayError(reason) else { return }
        showError(displayed)
    }

    /// Route a display error to the platform's error surface. iOS uses the
    /// shared `ConfigStore` banner; macOS overrides this to route through its
    /// global alert (`UiStore`).
    open func showError(_ error: DisplayError) {
        #if os(iOS)
            configStore.showError(error)
        #else
            preconditionFailure(
                "macOS AppService must override showError to route through UiStore"
            )
        #endif
    }

    public func startCommonSubscriptions() {
        commonSubscriptions.start(
            applyPlayback: { [weak self] values in
                guard let self else { return }
                playbackValues.apply(values)
                mediaControlService.applyMediaControlValues(
                    values.mediaControl,
                    appHandle: appHandle
                )
                applyPlatformPlaybackValues(values)
            },
            applyQueue: { [weak self] snapshot in
                self?.playbackValues.applyQueueSnapshot(snapshot)
                self?.mediaControlService
                    .updateCommandAvailability(
                        hasNext: snapshot.hasNext,
                        hasPrevious: snapshot.hasPrevious
                    )
            },
            onError: { [weak self] error in self?.showError(error) }
        )
    }

    open func applyPlatformPlaybackValues(_: BridgePlaybackValues) {}

    func applyQueueItemsAdded(_ count: UInt32) {
        playbackValues.applyQueueItemsAdded(count)
    }

    #if DEBUG
        public var stateForTesting: AppServiceTestState {
            testAccess.state
        }

        public var queueItemsAddedPublisherForTesting: AnyPublisher<Int, Never>
        {
            testAccess.queueItemsAddedPublisher
        }

    #endif
}

extension AppService {
    public func installSharedEnvironment<Content: View>(
        _ content: Content
    ) -> some View {
        content
            .environment(playbackStore)
            .environment(configStore)
            .environment(syncStatusStore)
            .environment(artworkLoadingStore)
            .environment(libraryStore)
            .environment(libraryProjectionStore)
            .environment(libraryListsStore)
            .environment(downloadStore)
            .environment(outboxStore)
            .environment(downloads)
            .environment(imageStore)
            .environment(library)
            .environment(playback)
            .environment(queue)
            .environment(sync)
            .environment(cast)
            .environment(castStore)
            .environment(mediaPaths)
    }
}

File: bae-windows/Services/AppService.cs
using System;
using Microsoft.UI.Dispatching;
using Microsoft.UI.Xaml.Controls;
using uniffi.bae_bridge;

namespace Bae.Windows;

/// <summary>
/// Root object for one open library — the Windows analog of BaeKit's
/// <c>AppService</c>. Built once around an already-open <see cref="SessionStore"/>,
/// it bundles the narrow domain services and the reactive stores that views and
/// stores consume instead of reaching for the session and <see cref="NativeBae"/>
/// directly. The session keeps its lifecycle role (open/subscribe/teardown); the
/// AppService is the thing built around it.
/// </summary>
internal sealed class AppService
{
    // Domain services: narrow, per-domain projections of the open handle that
    // views and stores read instead of the session + NativeBae directly.
    public LibraryService Library { get; }
    public MediaPathsService MediaPaths { get; }
    public PlaybackService Playback { get; }
    public QueueService Queue { get; }
    public DownloadsService Downloads { get; }
    public SyncService Sync { get; }
    public CastService Cast { get; }
    public SettingsService Settings { get; }
    public ImportService Import { get; }

    // The system media transport controls, bound to the window's HWND at
    // construction. One instance for the window's lifetime; a library switch
    // deactivates it rather than recreating it.
    public MediaControlService MediaControlService { get; }

    // The reactive stores. One instance per open library: a switch builds a
    // fresh AppService with fresh stores (mirroring macOS's AppService).
    public ShellStore ShellStore { get; }
    public PlaybackStore PlaybackStore { get; }
    public CastStore CastStore { get; }
    public SyncStatusStore SyncStatusStore { get; }
    public SettingsStore SettingsStore { get; }
    public ImportStore ImportStore { get; }
    public StorageStore StorageStore { get; }
    public TransferProgressStore TransferProgressStore { get; }
    public ProjectionRegistry ProjectionRegistry { get; }
    public UiEventRouter UiEventRouter { get; }
    public LibraryBrowserStore LibraryBrowserStore { get; }

    // windowHandle is invoked once here (MediaControlService binds the HWND at
    // construction); the caller (MainWindow) has the live window handle.
    public AppService(SessionStore session, DispatcherQueue dispatcher, Func<IntPtr> windowHandle)
        : this(
            session,
            dispatcher,
            windowHandle,
            LibraryService.FromSession(session),
            MediaPathsService.FromSession(session),
            PlaybackService.FromSession(session),
            QueueService.FromSession(session),
            DownloadsService.FromSession(session),
            SyncService.FromSession(session),
            CastService.FromSession(session),
            SettingsService.FromSession(session),
            ImportService.FromSession(session))
    {
    }

    // The domain services are injected so a scene can override them (see Stubbed);
    // the stores, transport controls, and router are the same for every
    // composition and are built here around the given session and services.
    private AppService(
        SessionStore session,
        DispatcherQueue dispatcher,
        Func<IntPtr> windowHandle,
        LibraryService library,
        MediaPathsService mediaPaths,
        PlaybackService playback,
        QueueService queue,
        DownloadsService downloads,
        SyncService sync,
        CastService cast,
        SettingsService settings,
        ImportService import)
    {
        Library = library;
        MediaPaths = mediaPaths;
        Playback = playback;
        Queue = queue;
        Downloads = downloads;
        Sync = sync;
        Cast = cast;
        Settings = settings;
        Import = import;

        ShellStore = new ShellStore();
        PlaybackStore = new PlaybackStore();
        CastStore = new CastStore(Cast);
        SyncStatusStore = new SyncStatusStore(Sync);
        // Built before the now-playing bar (in MainView): the bar reads the
        // remaining-time preference through the settings mirror, whose Current
        // stays null until a library seeds it.
        SettingsStore = new SettingsStore(Settings);
        MediaControlService = new MediaControlService(
            windowHandle(),
            dispatcher,
            Playback,
            image => MediaPaths.FetchLibraryImageBytes(image));
        ImportStore = new ImportStore(Import, ShowError, MediaControlService);
        ProjectionRegistry = new ProjectionRegistry();
        // In-flight release transfers (pin/unpin/manage/unmanage), driven by
        // core's ReleaseTransferProgress/Ended events the router routes, read by
        // the storage dialog rows and the album-detail storage band. Built before
        // the router so it can route into it, and shared with the stores below.
        TransferProgressStore = new TransferProgressStore();
        UiEventRouter = new UiEventRouter(
            PlaybackStore,
            ShowError,
            ProjectionRegistry,
            MediaControlService,
            ImportStore.HandlePreviewEvent,
            TransferProgressStore,
            CastStore);
        // The storage sheet's non-UI operations, shared by the storage dialog and
        // the album-detail storage band so both run transitions the same way.
        StorageStore = new StorageStore(Downloads, Sync, TransferProgressStore);
        LibraryBrowserStore = new LibraryBrowserStore(Library, MediaPaths, dispatcher);
    }

#if DEBUG
    /// <summary>A scene composition for the screenshot gallery: every domain
    /// service is a fail-loud stub except the caller-supplied <paramref name="library"/>,
    /// so a render that touches any unwired delegate crashes the capture loudly
    /// rather than drawing a lie. The stores are built around a handle-less
    /// <paramref name="session"/>; the shell renders its content off the stub
    /// library alone and never reaches a live handle.</summary>
    public static AppService Stubbed(
        SessionStore session,
        DispatcherQueue dispatcher,
        Func<IntPtr> windowHandle,
        LibraryService library) =>
        new(
            session,
            dispatcher,
            windowHandle,
            library,
            new MediaPathsService(),
            new PlaybackService(),
            new QueueService(),
            new DownloadsService(),
            new SyncService(),
            new CastService(),
            new SettingsService(),
            new ImportService());
#endif

    /// <summary>Route a caught error to the shell's error banner — the macOS
    /// AppService.showError analog, the single door feature code reaches for
    /// instead of writing the shell store's banner directly. Every caller surfaces
    /// an error, so the severity is fixed.</summary>
    public void ShowError(string title, string message) =>
        ShellStore.ShowBanner(InfoBarSeverity.Error, title, message);

    /// <summary>Report that a host UI screen opened, as a typed telemetry event
    /// through the process-lifetime sink. Infallible; the core owns every other
    /// event. The macOS AppService.reportScreen analog.</summary>
    public void ReportScreen(BridgeScreen screen) =>
        NativeBae.ReportScreen(BaeDiagnostics.Handle, screen);
}
